### ISSUE TITLE
test_reentry.py & py3 -> 5 failures from 8 tests

### ISSUE THREAD

**Issue by jaraco** [jaraco/jaraco.vcs#14]
Originally reported by: **BitBucket: [idella5](http://bitbucket.org/idella5), GitHub: [idella5](http://github.com/idella5)**

---

hgtools-3.0.2-python3_3 $ PYTHONPATH=build/lib/ py.test-python3.3 build/lib//hgtools/tests/

yields 

```
build/lib/hgtools/tests/__init__.py .
build/lib/hgtools/tests/test_managers.py .............
build/lib/hgtools/tests/test_reentry.py .FFFFF..
build/lib/hgtools/tests/test_versioning.py ......

============================================================== FAILURES ===============================================================
____________________________________________________ TestReEntry.test_hello_world _____________________________________________________

self = <hgtools.tests.test_reentry.TestReEntry object at 0x7fb3ea095610>

    def test_hello_world(self):
            with reentry.in_process_context(['hello-world.py']) as proc:
>                   hello_world()

build/lib/hgtools/tests/test_reentry.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def hello_world():
>       print("hello world")
E    TypeError: 'str' does not support the buffer interface

build/lib/hgtools/tests/test_reentry.py:10: TypeError
_______________________________________________ TestReEntry.test_main_with_system_exit ________________________________________________

self = <hgtools.tests.test_reentry.TestReEntry object at 0x7fb3ea0ac550>

    def test_main_with_system_exit(self):
            with reentry.in_process_context(['exit-zero']) as proc:
                    exit_zero()
            assert proc.returncode == 0
>           assert proc.stdio.stdout.getvalue() == ''
E           assert b'' == ''
E            +  where b'' = <built-in method getvalue of _io.BytesIO object at 0x7fb3ecdaa530>()
E            +    where <built-in method getvalue of _io.BytesIO object at 0x7fb3ecdaa530> = <_io.BytesIO object at 0x7fb3ecdaa530>.getvalue
E            +      where <_io.BytesIO object at 0x7fb3ecdaa530> = SavedIO(stdout=<_io.BytesIO object at 0x7fb3ecdaa530>, stderr=<_io.BytesIO object at 0x7fb3ecdaa668>).stdout
E            +        where SavedIO(stdout=<_io.BytesIO object at 0x7fb3ecdaa530>, stderr=<_io.BytesIO object at 0x7fb3ecdaa668>) = <hgtools.managers.reentry.ProcessResult object at 0x7fb3ea0acfd0>.stdio

build/lib/hgtools/tests/test_reentry.py:38: AssertionError
_____________________________________________ TestReEntry.test_main_with_system_exit_one ______________________________________________

self = <hgtools.tests.test_reentry.TestReEntry object at 0x7fb3ea0aa050>

    def test_main_with_system_exit_one(self):
            with reentry.in_process_context(['exit-one']) as proc:
                    exit_one()
            assert proc.returncode == 1
>           assert proc.stdio.stdout.getvalue() == ''
E           assert b'' == ''
E            +  where b'' = <built-in method getvalue of _io.BytesIO object at 0x7fb3ecdaa738>()
E            +    where <built-in method getvalue of _io.BytesIO object at 0x7fb3ecdaa738> = <_io.BytesIO object at 0x7fb3ecdaa738>.getvalue
E            +      where <_io.BytesIO object at 0x7fb3ecdaa738> = SavedIO(stdout=<_io.BytesIO object at 0x7fb3ecdaa738>, stderr=<_io.BytesIO object at 0x7fb3ecdaaa78>).stdout
E            +        where SavedIO(stdout=<_io.BytesIO object at 0x7fb3ecdaa738>, stderr=<_io.BytesIO object at 0x7fb3ecdaaa78>) = <hgtools.managers.reentry.ProcessResult object at 0x7fb3ea0aaa50>.stdio

build/lib/hgtools/tests/test_reentry.py:44: AssertionError
____________________________________________ TestReEntry.test_main_with_system_exit_string ____________________________________________

self = <hgtools.tests.test_reentry.TestReEntry object at 0x7fb3ea512f50>

    def test_main_with_system_exit_string(self):
            with reentry.in_process_context(['exit-string']) as proc:
                    exit_string()
            assert proc.returncode == 1
>           assert proc.stdio.stdout.getvalue() == ''
E           assert b'' == ''
E            +  where b'' = <built-in method getvalue of _io.BytesIO object at 0x7fb3ecdaa6d0>()
E            +    where <built-in method getvalue of _io.BytesIO object at 0x7fb3ecdaa6d0> = <_io.BytesIO object at 0x7fb3ecdaa6d0>.getvalue
E            +      where <_io.BytesIO object at 0x7fb3ecdaa6d0> = SavedIO(stdout=<_io.BytesIO object at 0x7fb3ecdaa6d0>, stderr=<_io.BytesIO object at 0x7fb3ecdaa870>).stdout
E            +        where SavedIO(stdout=<_io.BytesIO object at 0x7fb3ecdaa6d0>, stderr=<_io.BytesIO object at 0x7fb3ecdaa870>) = <hgtools.managers.reentry.ProcessResult object at 0x7fb3ea512a90>.stdio

build/lib/hgtools/tests/test_reentry.py:50: AssertionError
_____________________________________________________ TestReEntry.test_echo_args ______________________________________________________

self = <hgtools.tests.test_reentry.TestReEntry object at 0x7fb3ea0c1690>

    def test_echo_args(self):
            with reentry.in_process_context(['echo', 'foo', 'bar']) as proc:
>                   echo()

build/lib/hgtools/tests/test_reentry.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def echo():
>       print("args are", sys.argv)
E    TypeError: 'str' does not support the buffer interface

build/lib/hgtools/tests/test_reentry.py:22: TypeError
================================================= 5 failed, 23 passed in 7.74 seconds =================================================
```

ditto py3.2 3.2

hgtools-3.0.2

---
- Bitbucket: https://bitbucket.org/jaraco/hgtools/issue/14

**Comment by jaraco** [comment]
_Original comment by_: **Matthew Thode (BitBucket: [prometheanfire](http://bitbucket.org/prometheanfire), GitHub: [prometheanfire](http://github.com/prometheanfire))**

---

Here, have a downstream bug report as well :D

https://bugs.gentoo.org/show_bug.cgi?id=482492

**Comment by jaraco** [comment]
_Original comment by_: **Jason R. Coombs (BitBucket: [jaraco](http://bitbucket.org/jaraco), GitHub: [jaraco](http://github.com/jaraco))**

---

Use StringIO on Python 3 in capture_stdio. Fixes failing tests reported in #14. Fixes #14.

**Comment by jaraco** [comment]
_Original comment by_: **Jason R. Coombs (BitBucket: [jaraco](http://bitbucket.org/jaraco), GitHub: [jaraco](http://github.com/jaraco))**

---

Released with hgtools 4.0.
